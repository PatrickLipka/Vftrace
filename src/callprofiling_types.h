#ifndef CALLPROFILING_TYPES_H
#define CALLPROFILING_TYPES_H

typedef struct {
   // number of calls
   long long calls;
   // cycles spend in the function (including subfunctions)
   long long cycles;
   // time spend in the function (including subfunctions)
   long long time_usec;
   // time spend in the function (excluding subfunctions)
   // computed during final stack update
   long long time_excl_usec;
} callProfile_t;

#endif
