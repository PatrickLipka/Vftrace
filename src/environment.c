#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <limits.h>

#include "self_profile.h"
#include "environment.h"
#include "regular_expressions.h"
#include "range_expand.h"
#include "misc_utils.h"

env_var_t *vftr_get_env_var_ptr_by_idx(environment_t *environment_ptr, int idx) {
   env_var_t *environment_arr = (env_var_t*) environment_ptr;
   if (idx >= 0 && idx < environment_ptr->nenv_vars) {
      return environment_arr+idx;
   } else {
      return NULL;
   }
}

env_var_t vftr_read_env_int(char *env_name, int default_val) {
   env_var_t env_var;
   env_var.value_kind = env_int;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      env_var.value.int_val = atoi(env_var.value_string);
      env_var.set = true;
   } else {
      env_var.value.int_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_long(char *env_name, long default_val) {
   env_var_t env_var;
   env_var.value_kind = env_long;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      env_var.value.long_val = atol(env_var.value_string);
      env_var.set = true;
   } else {
      env_var.value.long_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_longlong(char *env_name, long long default_val) {
   env_var_t env_var;
   env_var.value_kind = env_longlong;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      env_var.value.longlong_val = atoll(env_var.value_string);
      env_var.set = true;
   } else {
      env_var.value.longlong_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_float(char *env_name, float default_val) {
   env_var_t env_var;
   env_var.value_kind = env_float;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      sscanf(env_var.value_string, "%f", &(env_var.value.float_val));
      env_var.set = true;
   } else {
      env_var.value.float_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_double(char *env_name, double default_val) {
   env_var_t env_var;
   env_var.value_kind = env_double;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      sscanf(env_var.value_string, "%lf", &(env_var.value.double_val));
      env_var.set = true;
   } else {
      env_var.value.double_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_bool(char *env_name, bool default_val) {
   env_var_t env_var;
   env_var.value_kind = env_bool;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      char *value_string = strdup(env_var.value_string);
      char *s = value_string;
      while (*s != '\0') {
         *s = tolower(*s);
         s++;
      }
      if (!strcmp(value_string, "1") ||
          !strcmp(value_string, "yes") ||
          !strcmp(value_string, "on")) {
         env_var.value.bool_val = true;
      } else {
         env_var.value.bool_val = false;
      }
      env_var.set = true;
      free(value_string);
   } else {
      env_var.value.bool_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_string(char *env_name, char *default_val) {
   env_var_t env_var;
   env_var.value_kind = env_string;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      env_var.value.string_val = env_var.value_string;
      env_var.set = true;
   } else {
      env_var.value.string_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

env_var_t vftr_read_env_regex(char *env_name, regex_t *default_val) {
   env_var_t env_var;
   env_var.value_kind = env_regex;
   env_var.name = env_name;
   env_var.value_string = getenv(env_name);
   if (env_var.value_string != NULL) {
      env_var.value.regex_val = vftr_compile_regexp(env_var.value_string);
      env_var.set = true;
   } else {
      env_var.value.regex_val = default_val;
      env_var.set = false;
   }
   return env_var;
}

void vftr_print_env_var(FILE *fp, env_var_t env_var) {
   switch (env_var.value_kind) {
      case env_int:
         fprintf(fp, "%s: %d%s\n",
                 env_var.name,
                 env_var.value.int_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_long:
         fprintf(fp, "%s: %ld%s\n",
                 env_var.name,
                 env_var.value.long_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_longlong:
         fprintf(fp, "%s: %lld%s\n",
                 env_var.name,
                 env_var.value.longlong_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_float:
         fprintf(fp, "%s: %f%s\n",
                 env_var.name,
                 env_var.value.float_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_double:
         fprintf(fp, "%s: %lf%s\n",
                 env_var.name,
                 env_var.value.double_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_bool:
         fprintf(fp, "%s: %s%s\n",
                 env_var.name,
                 env_var.value.bool_val ? "true" : "false",
                 env_var.set ? "" : "(default)");
         break;
      case env_string:
         fprintf(fp, "%s: %s%s\n",
                 env_var.name,
                 env_var.value.string_val,
                 env_var.set ? "" : "(default)");
         break;
      case env_regex:
         fprintf(fp, "%s: %s%s\n",
                 env_var.name,
                 env_var.value_string,
                 env_var.set ? "" : "(default)");
         break;
      case env_none:
      default:
         fprintf(fp, "Invalid environment variable type\n");
         break;
   }
}

void vftr_print_environment(FILE *fp, environment_t environment) {
   if (environment.valid && environment.print_environment.value.bool_val) {
      for (int ienv=0; ienv<environment.nenv_vars; ienv++) {
         vftr_print_env_var(fp,
            *vftr_get_env_var_ptr_by_idx(&environment, ienv));
      }
   } else {
      fprintf(fp, "Environment is invalid!\n"
                  "Not read yet, or already freed!\n");
   }
}

environment_t vftr_read_environment() {
   SELF_PROFILE_START_FUNCTION;
   environment_t environment;
   environment.vftrace_off = vftr_read_env_bool("VFTR_OFF", false);
   environment.do_sampling = vftr_read_env_bool("VFTR_SAMPLING", false);
   environment.output_directory = vftr_read_env_string("VFTR_OUT_DIRECTORY", ".");
   environment.logfile_basename = vftr_read_env_string("VFTR_LOGFILE_BASENAME", NULL);
   environment.logfile_for_ranks = vftr_read_env_string("VFTR_LOGFILE_FOR_RANKS", "none");
   environment.ranks_in_mpi_profile = vftr_read_env_string("VFTR_RANKS_IN_MPI_PROFILE", "all");
   environment.sampletime = vftr_read_env_double("VFTR_SAMPLETIME", 0.005);
   environment.mpi_log = vftr_read_env_bool("VFTR_MPI_LOG", false);
   environment.mpi_show_sync_time = vftr_read_env_bool("VFTR_MPI_SHOW_SYNC_TIME", false);
   environment.vfd_bufsize = vftr_read_env_int("VFTR_VFD_BUFSIZE", 8);
   environment.include_only_regex = vftr_read_env_regex("VFTR_INCLUDE_ONLY", NULL);
   environment.scenario_file = vftr_read_env_string("VFTR_SCENARIO_FILE", NULL);
   environment.preciseregex = vftr_read_env_regex("VFTR_PRECISE", NULL);
   environment.print_stack_profile = vftr_read_env_regex("VFTR_PRINT_STACK_PROFILE", NULL);
   environment.print_stacks_for = vftr_read_env_string("VFTR_PRINT_STACKS_FOR", NULL);
   environment.strip_module_names = vftr_read_env_bool("VFTR_STRIP_MODULE_NAMES", false);
   environment.sort_profile_table = vftr_read_env_string("VFTR_SORT_PROFILE_TABLE", "TIME_EXCL");
   environment.sort_mpi_table = vftr_read_env_string("VFTR_SORT_MPI_TABLE", "NONE");
   environment.show_overhead = vftr_read_env_bool("VFTR_SHOW_FUNCTION_OVERHEAD", false);
   environment.show_calltime_imbalances = vftr_read_env_bool("VFTR_SHOW_CALLTIME_IMBALANCES", false);
   environment.group_functions_by_name = vftr_read_env_bool("VFTR_GROUP_FUNCTIONS_BY_NAME", false);
   environment.print_environment = vftr_read_env_bool("VFTR_PRINT_ENVIRONMENT", true);
   environment.callpath_in_profile = vftr_read_env_bool("VFTR_CALLPATH_IN_PROFILE", false);
   environment.callpath_in_mpi_profile = vftr_read_env_bool("VFTR_CALLPATH_IN_MPI_PROFILE", false);
   environment.demangle_cxx = vftr_read_env_bool("VFTR_DEMANGLE_CXX", false);
   environment.nenv_vars = 25;
   environment.valid = true;
   SELF_PROFILE_END_FUNCTION;
   return environment;
}

// Loop over all Vftrace environment variables. When LD is zero, we have an exact match
// and we exit the subroutine to save time.
void vftr_env_var_find_best_match(environment_t *environment_ptr,
                                  char *var_name, int *best_ld,
                                  int *best_idx) {
   *best_ld = INT_MAX;
   *best_idx = -1;
   for (int ivar=0; ivar<environment_ptr->nenv_vars; ivar++) {
      env_var_t *env_var = vftr_get_env_var_ptr_by_idx(environment_ptr, ivar);
      int ld = vftr_levenshtein_distance(var_name,
                                         env_var->name);
      if (ld < *best_ld) {
         *best_ld = ld;
         *best_idx = ivar;
      }
      if (ld == 0) {
         return;
      }
   }
}

// There might be mistyped Vftrace environment variables. Loop over all existing env variables,
// check if they match a Vftrace variable, and make an alternative suggestion if it is possibly mistyped.
void vftr_check_env_names(FILE *fp, environment_t *environment_ptr) {
   SELF_PROFILE_START_FUNCTION;
   extern char **environ;
   char **s = environ;
   for (; *s; s++) {
      if (strstr(*s, "VFTR_")) {
         // There has to be strdup of s, because strtok modifies the first argument. This
         // points to the external environ, which is better not touched!
         char *tmpstr = strdup(*s);
         char *var_name = strtok(tmpstr, "=");
         int best_ld, best_idx;
         vftr_env_var_find_best_match(environment_ptr, var_name, &best_ld, &best_idx);
         // best_ld == 0 -> Exact match.
         if (best_ld > 0)  {
            env_var_t *env_var = vftr_get_env_var_ptr_by_idx(environment_ptr, best_idx);
            fprintf(fp, "Vftrace environment variable %s not known. Do you mean %s?\n",
                    var_name, env_var->name);
         }
         free(tmpstr);
      }
   }
   SELF_PROFILE_END_FUNCTION;
}

void vftr_env_var_free(env_var_t *env_var_ptr) {
   env_var_t env_var = *env_var_ptr;
   if (env_var_ptr != NULL) {
      switch (env_var.value_kind) {
         case env_int:
            break;
         case env_long:
            break;
         case env_longlong:
            break;
         case env_float:
            break;
         case env_double:
            break;
         case env_bool:
            break;
         case env_string:
            break;
         case env_regex:
            if (env_var.set) {
               regfree(env_var.value.regex_val);
               free(env_var.value.regex_val);
               env_var.value.regex_val = NULL;
            }
            break;
         default:
            break;
      }
   }
}

void vftr_environment_free(environment_t *environment_ptr) {
   SELF_PROFILE_START_FUNCTION;
   environment_t environment = *environment_ptr;
   if (environment.valid) {
      environment.valid = false;
      for (int ienv=0; ienv<environment_ptr->nenv_vars; ienv++) {
         vftr_env_var_free(vftr_get_env_var_ptr_by_idx(environment_ptr, ienv));
      }
   }
   SELF_PROFILE_END_FUNCTION;
}

// Attempt to check the user supplied environment values for soundness
void vftr_environment_assert_vftrace_off(FILE *fp, env_var_t vftrace_off) {
   (void) fp;
   (void) vftrace_off;
}

void vftr_environment_assert_do_sampling(FILE *fp, env_var_t do_sampling) {
   (void) fp;
   (void) do_sampling;
}

void vftr_environment_assert_output_directory(FILE *fp, env_var_t output_directory) {
   (void) fp;
   (void) output_directory;
}

void vftr_environment_assert_logfile_basename(FILE *fp, env_var_t logfile_basename) {
   (void) fp;
   (void) logfile_basename;
}

void vftr_environment_assert_logfile_for_ranks(FILE *fp, env_var_t logfile_for_ranks) {
   char *rangelist = logfile_for_ranks.value.string_val;
   if (!strcmp(rangelist, "all")) {
      ;
   } else if (!strcmp(rangelist, "none")) {
      ;
   } else {
      int nvals = 0;
      int *exp_list = vftr_expand_rangelist(rangelist, &nvals);
      if (nvals == 0 || exp_list == NULL) {
         fprintf(fp, "Warning: Unable to properly parse given list for \"%s\".\n",
                 rangelist);
      }
      for (int i=0; i<nvals; i++) {
         if (exp_list[i] < 0) {
            fprintf(fp, "Warning: %s=%s results in negative rank values.",
                    logfile_for_ranks.name, rangelist);
         }
      }
      free(exp_list);
   }
}

void vftr_environment_assert_ranks_in_mpi_profile(FILE *fp,
                                                  env_var_t ranks_in_mpi_profile) {
   char *rangelist = ranks_in_mpi_profile.value.string_val;
   if (strcmp(rangelist, "all")) {
      int nvals = 0;
      int *exp_list = vftr_expand_rangelist(rangelist, &nvals);
      if (nvals == 0 || exp_list == NULL) {
         fprintf(fp, "Warning: Unable to properly parse given list for \"%s\".\n",
                 rangelist);
      }
      for (int i=0; i<nvals; i++) {
         if (exp_list[i] < 0) {
            fprintf(fp, "Warning: %s=%s results in negative rank values.",
                    ranks_in_mpi_profile.name, rangelist);
         }
      }
      free(exp_list);
   }
}

void vftr_environment_assert_sampletime(FILE *fp, env_var_t sampletime) {
   if (sampletime.value.double_val <= 0.0) {
      fprintf(fp, "Warning: \"%s\" with a value smaller or equal to 0 "
              "will sample every event.\n",
              sampletime.name);
   }
}

void vftr_environment_assert_mpi_log(FILE *fp, env_var_t mpi_log) {
   (void) fp;
   (void) mpi_log;
}

void vftr_environment_assert_mpi_show_sync_time(FILE *fp,
                                                env_var_t mpi_show_sync_time) {
   (void) fp;
   (void) mpi_show_sync_time;
}

void vftr_environment_assert_vfd_bufsize(FILE *fp, env_var_t vfd_bufsize) {
   if (vfd_bufsize.value.int_val <= 0) {
      fprintf(fp, "Warning: \"%s\" needs to be > 0, but is %d\n",
              vfd_bufsize.name, vfd_bufsize.value.int_val);
   }
}

void vftr_environment_assert_include_only_regex(FILE *fp, env_var_t include_only_regex) {
   (void) fp;
   (void) include_only_regex;
}

void vftr_environment_assert_scenario_file(FILE *fp, env_var_t scenario_file) {
   if (scenario_file.set) {
      char *scn_file = scenario_file.value.string_val;
      if (strlen(scn_file) == 0) {
         fprintf(fp, "Warning: %s set, but to an empty value.\n",
                 scenario_file.name);
      } else {
         if (access(scn_file, F_OK) != 0) {
            fprintf(fp, "Warning: %s is set to %s, but file could not be found\n",
                    scenario_file.name, scenario_file.value.string_val);
         }
      }
   }
}

void vftr_environment_assert_preciseregex(FILE *fp, env_var_t preciseregex) {
   (void) fp;
   (void) preciseregex;
}

void vftr_environment_assert_print_stack_profile(FILE *fp,
                                                 env_var_t print_stack_profile) {
   (void) fp;
   (void) print_stack_profile;
}

void vftr_environment_assert_print_stacks_for(FILE *fp, env_var_t print_stacks_for) {
   (void) fp;
   (void) print_stacks_for;
}

void vftr_environment_assert_strip_module_names(FILE *fp, env_var_t strip_module_names) {
   (void) fp;
   (void) strip_module_names;
}

void vftr_environment_assert_sort_profile_table(FILE *fp, env_var_t sort_profile_table) {
   char *sort_str = sort_profile_table.value.string_val;
   int nvalid_str = 6;
   char *valid_str[] = {"TIME_EXCL", "TIME_INCL", "CALLS",
                        "STACK_ID", "OVERHEAD", "NONE"};
   bool valid = false;
   for (int istr=0; istr<nvalid_str; istr++) {
      valid = valid || !strcmp(sort_str, valid_str[istr]);
   }
   if (!valid) {
      fprintf(fp, "Warning: %s was set to \"%s\", but only valid options are:",
              sort_profile_table.name, sort_str);
      for (int istr=0; istr<nvalid_str-1; istr++) {
         fprintf(fp, " \"%s\",", valid_str[istr]);
      }
      fprintf(fp, " and \"%s\".\n", valid_str[nvalid_str-1]);
   }
}

void vftr_environment_assert_sort_mpi_table(FILE *fp, env_var_t sort_mpi_table) {
   char *sort_str = sort_mpi_table.value.string_val;
   int nvalid_str = 8;
   char *valid_str[] = {"MESSAGES",
                        "SEND_SIZE", "RECV_SIZE",
                        "SEND_BW", "SEND_BW",
                        "COMM_TIME", "STACK_ID",
                        "NONE"};
   bool valid = false;
   for (int istr=0; istr<nvalid_str; istr++) {
      valid = valid || !strcmp(sort_str, valid_str[istr]);
   }
   if (!valid) {
      fprintf(fp, "Warning: %s was set to \"%s\", but only valid options are:",
              sort_mpi_table.name, sort_str);
      for (int istr=0; istr<nvalid_str-1; istr++) {
         fprintf(fp, " \"%s\",", valid_str[istr]);
      }
      fprintf(fp, " and \"%s\".\n", valid_str[nvalid_str-1]);
   }
}

void vftr_environment_assert_show_overhead(FILE *fp, env_var_t show_overhead) {
   (void) fp;
   (void) show_overhead;
}

void vftr_environment_assert_show_calltime_imbalances(FILE *fp, env_var_t show_calltime_imbalances) {
   (void) fp;
   (void) show_calltime_imbalances;
}

void vftr_environment_assert_group_functions_by_name(FILE* fp, env_var_t group_functions_by_name) {
   (void) fp;
   (void) group_functions_by_name;
}

void vftr_environment_assert_print_environment(FILE *fp, env_var_t print_environment) {
   (void) fp;
   (void) print_environment;
}

void vftr_environment_assert_callpath_in_profile(FILE *fp,
                                                 env_var_t callpath_in_profile) {
   (void) fp;
   (void) callpath_in_profile;
}

void vftr_environment_assert_callpath_in_mpi_profile(FILE *fp,
                                                     env_var_t callpath_in_mpi_profile) {
   (void) fp;
   (void) callpath_in_mpi_profile;
}

void vftr_environment_assert_demangle_cxx(FILE *fp, env_var_t demangle_cxx) {
   (void) fp;
   (void) demangle_cxx;
}

void vftr_environment_assert(FILE *fp, environment_t environment) {
   SELF_PROFILE_START_FUNCTION;
   vftr_environment_assert_vftrace_off(fp, environment.vftrace_off);
   vftr_environment_assert_do_sampling(fp, environment.do_sampling);
   vftr_environment_assert_output_directory(fp, environment.output_directory);
   vftr_environment_assert_logfile_basename(fp, environment.logfile_basename);
   vftr_environment_assert_logfile_for_ranks(fp, environment.logfile_for_ranks);
   vftr_environment_assert_ranks_in_mpi_profile(fp, environment.ranks_in_mpi_profile);
   vftr_environment_assert_sampletime(fp, environment.sampletime);
   vftr_environment_assert_mpi_log(fp, environment.mpi_log);
   vftr_environment_assert_mpi_show_sync_time(fp, environment.mpi_show_sync_time);
   vftr_environment_assert_vfd_bufsize(fp, environment.vfd_bufsize);
   vftr_environment_assert_include_only_regex(fp, environment.include_only_regex);
   vftr_environment_assert_scenario_file(fp, environment.scenario_file);
   vftr_environment_assert_preciseregex(fp, environment.preciseregex);
   vftr_environment_assert_print_stack_profile(fp, environment.print_stack_profile);
   vftr_environment_assert_print_stacks_for(fp, environment.print_stacks_for);
   vftr_environment_assert_strip_module_names(fp, environment.strip_module_names);
   vftr_environment_assert_sort_profile_table(fp, environment.sort_profile_table);
   vftr_environment_assert_sort_mpi_table(fp, environment.sort_mpi_table);
   vftr_environment_assert_show_overhead(fp, environment.show_overhead);
   vftr_environment_assert_show_calltime_imbalances(fp, environment.show_calltime_imbalances);
   vftr_environment_assert_group_functions_by_name(fp, environment.group_functions_by_name);
   vftr_environment_assert_print_environment(fp, environment.print_environment);
   vftr_environment_assert_callpath_in_profile(fp, environment.callpath_in_profile);
   vftr_environment_assert_callpath_in_mpi_profile(fp, environment.callpath_in_mpi_profile);
   vftr_environment_assert_demangle_cxx(fp, environment.demangle_cxx);
   SELF_PROFILE_END_FUNCTION;
}
