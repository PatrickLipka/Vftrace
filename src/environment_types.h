#ifndef ENVIRONMENT_TYPES_H
#define ENVIRONMENT_TYPES_H

#include <stdbool.h>

#include <regex.h>

typedef enum {
   env_none,
   env_int,
   env_long,
   env_longlong,
   env_float,
   env_double,
   env_bool,
   env_string,
   env_regex
} env_var_kind;

typedef union {
   int int_val;
   long long_val;
   long long longlong_val;
   float float_val;
   double double_val;
   bool bool_val;
   char *string_val;
   regex_t *regex_val;
} env_var_value;

typedef struct {
   env_var_kind value_kind;
   env_var_value value;
   char *value_string;
   char *name;
   bool set;
} env_var_t;

typedef struct {
   // in order to access the environment variables
   // like an array by index, all non-environment variables
   // need to be at the end of the struct!
   env_var_t vftrace_off;
   env_var_t do_sampling;
   env_var_t output_directory;
   env_var_t logfile_basename;
   env_var_t logfile_for_ranks;
   env_var_t ranks_in_mpi_profile;
   env_var_t sampletime;
   env_var_t mpi_log;
   env_var_t mpi_show_sync_time;
   env_var_t vfd_bufsize;
   env_var_t include_only_regex;
   env_var_t scenario_file;
   env_var_t preciseregex;
   env_var_t print_stack_profile;
   env_var_t print_stacks_for;
   env_var_t strip_module_names;
   env_var_t sort_profile_table;
   env_var_t sort_mpi_table;
   env_var_t show_overhead;
   env_var_t show_calltime_imbalances;
   env_var_t print_environment;
   env_var_t callpath_in_profile;
   env_var_t callpath_in_mpi_profile;
   env_var_t demangle_cxx;
   bool valid;
   int nenv_vars;
} environment_t;

#endif
