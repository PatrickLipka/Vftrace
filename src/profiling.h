#ifndef PROFILING_H
#define PROFILING_H

#include <stdlib.h>

#include "thread_types.h"
#include "profiling_types.h"

int vftr_new_profile(int threadID, profilelist_t *profilelist_ptr);

void vftr_profile_free(profile_t* profiles_ptr, int profID);

profilelist_t vftr_new_profilelist();

void vftr_profilelist_free(profilelist_t *profilelist_ptr);

profile_t *vftr_get_my_profile(stack_t *stack, thread_t *thread);

#endif
