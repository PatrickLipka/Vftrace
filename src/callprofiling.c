#include <stdio.h>
#include <stdbool.h>

#include "callprofiling_types.h"
#include "stack_types.h"

callProfile_t vftr_new_callprofiling() {
   callProfile_t prof;
   prof.calls = 0ll;
   prof.cycles = 0ll;
   prof.time_usec = 0ll;
   prof.time_excl_usec = 0ll;
   return prof;
}

void vftr_accumulate_callprofiling(callProfile_t *prof,
                                   int calls,
                                   long long cycles,
                                   long long time_usec) {
   prof->calls += calls;
   prof->cycles += cycles;
   prof->time_usec += time_usec;
}

void vftr_update_stacks_exclusive_time(int nstacks, stack_t *stacks) {
//TODO    // exclusive time for init is 0, therefore it does not need to be computed.
//TODO    for (int istack=1; istack<nstacks; istack++) {
//TODO       stack_t *mystack = stacks + istack;
//TODO       long long exclusive_time = mystack->profiling.callProf.time_usec;
//TODO       // subtract the time spent in the callees
//TODO       for (int icallee=0; icallee<mystack->ncallees; icallee++) {
//TODO          int calleeID = mystack->callees[icallee];
//TODO          exclusive_time -= stacks[calleeID].profiling.callProf.time_usec;
//TODO       }
//TODO       mystack->profiling.callProf.time_excl_usec = exclusive_time;
//TODO    }
}

void vftr_callprofiling_free(callProfile_t *callprof_ptr) {
   (void) callprof_ptr;
}

void vftr_print_callprofiling(FILE *fp, callProfile_t callprof) {
   fprintf(fp, "calls: %lld, cycles: %lld, time(incl/excl): %lld/%lld\n",
           callprof.calls, callprof.cycles,
           callprof.time_usec, callprof.time_excl_usec);
}
