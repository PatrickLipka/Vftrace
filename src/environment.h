#ifndef ENVIRONMENT_H
#define ENVIRONMENT_H

#include <stdio.h>

#include "environment_types.h"

env_var_t *vftr_get_env_var_ptr_by_idx(environment_t *environment_ptr, int idx);

void vftr_print_environment(FILE *fp, environment_t environment);

environment_t vftr_read_environment();

void vftr_check_env_names(FILE *fp, environment_t *environment_ptr);

void vftr_environment_free(environment_t *environment_ptr);

void vftr_environment_assert(FILE *fp, environment_t environment);

#endif
