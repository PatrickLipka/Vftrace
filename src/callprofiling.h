#ifndef CALLPROFILING_H
#define CALLPROFILING_H

#include <stdio.h>

#include "callprofiling_types.h"
#include "stack_types.h"

#include "callprofiling.h"

callProfile_t vftr_new_callprofiling();

void vftr_accumulate_callprofiling(callProfile_t *prof,
                                   int calls,
                                   long long cycles,
                                   long long time_usec);

void vftr_callprofiling_free(callProfile_t *callprof_ptr);

void vftr_update_stacks_exclusive_time(int nstacks, stack_t *stacks);

void vftr_print_callprofiling(FILE *fp, callProfile_t callprof);

#endif
