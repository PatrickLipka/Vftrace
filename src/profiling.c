#include <stdlib.h>

#include "vftrace_state.h"
#include "thread_types.h"
#include "profiling_types.h"
#include "realloc_consts.h"

#include "callprofiling.h"
#include "overheadprofiling.h"

void vftr_profilelist_realloc(profilelist_t *profilelist_ptr) {
   profilelist_t profilelist = *profilelist_ptr;
   while (profilelist.nprofiles > profilelist.maxprofiles) {
      int maxprofiles = profilelist.maxprofiles*vftr_realloc_rate+vftr_realloc_add;
      profilelist.profiles = (profile_t*)
         realloc(profilelist.profiles, maxprofiles*sizeof(profile_t));
      profilelist.maxprofiles = maxprofiles;
   }
   *profilelist_ptr = profilelist;
}

int vftr_new_profile(int threadID, profilelist_t *profilelist_ptr) {
   int profID = profilelist_ptr->nprofiles;
   profilelist_ptr->nprofiles++;
   vftr_profilelist_realloc(profilelist_ptr);

   // insert the new profile in such a way that they are sorted by their threadIDs
   for (int iprof=(profID-1); iprof>=0; iprof--) {
      if (threadID < profilelist_ptr->profiles[iprof].threadID) {
         profilelist_ptr->profiles[iprof+1] = profilelist_ptr->profiles[iprof];
         profID--;
      } else {
         break;
      }
   }

   profile_t *profile = profilelist_ptr->profiles+profID;
   profile->threadID = threadID;
   profile->callProf = vftr_new_callprofiling();
   profile->overheadProf = vftr_new_overheadprofiling();
   // TODO: Add other profiles

   return profID;
}

void vftr_profile_free(profile_t* profiles_ptr, int profID) {
   profile_t *profile_ptr = profiles_ptr+profID;
   vftr_callprofiling_free(&(profile_ptr->callProf));
   vftr_overheadprofiling_free(&(profile_ptr->overheadProf));
   // TODO: add other profiles
}

profilelist_t vftr_new_profilelist() {
   profilelist_t profilelist;
   profilelist.nprofiles = 0;
   profilelist.maxprofiles = 0;
   profilelist.profiles = NULL;
   return profilelist;
}

void vftr_profilelist_free(profilelist_t *profilelist_ptr) {
   profilelist_t profilelist = *profilelist_ptr;
   if (profilelist.nprofiles > 0) {
      for (int iprof=0; iprof<profilelist.nprofiles; iprof++) {
         vftr_profile_free(profilelist.profiles, iprof);
      }
      free(profilelist.profiles);
      profilelist.profiles = NULL;
      profilelist.nprofiles = 0;
      profilelist.maxprofiles = 0;
   }
   *profilelist_ptr = profilelist;
}

profile_t *vftr_get_my_profile(stack_t *stack,
                               thread_t *thread) {
   profilelist_t *profilelist_ptr = &(stack->profiling);
   // search for the profile matrhing the threadID
   // TODO: binary search?
   int profID = -1;
   for (int iprof=0; iprof<profilelist_ptr->nprofiles; iprof++) {
      profile_t *prof = profilelist_ptr->profiles+iprof;
      if (thread->threadID == prof->threadID) {
         profID = iprof;
         break;
      }
   }
   // if no matching profile is found create one
   // and update the profID
   if (profID == -1) {
      profID = vftr_new_profile(thread->threadID, profilelist_ptr);
   }

   return profilelist_ptr->profiles+profID;
}
