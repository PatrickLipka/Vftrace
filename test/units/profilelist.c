#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "environment_types.h"
#include "environment.h"
#include "stack_types.h"
#include "stacks.h"
#include "profiling_types.h"
#include "profiling.h"
#include "callprofiling_types.h"
#include "callprofiling.h"

#ifdef _MPI
#include <mpi.h>
#endif

int main(int argc, char **argv) {
   INIT_SELF_PROF_VFTRACE;
#if defined(_MPI)
   PMPI_Init(&argc, &argv);
#else
   (void) argc;
   (void) argv;
#endif

   environment_t environment;
   environment = vftr_read_environment();

   // build stacktree
   stacktree_t stacktree = vftr_new_stacktree();
   int nthreads = 10;
   for (int ithread=0; ithread<nthreads; ithread++) {
      vftr_new_profile_in_list(ithread, &(stacktree.stacks[0].profiling));
      profile_t *profile = stacktree.stacks[0].profiling.profiles+ithread;
      vftr_accumulate_callprofiling(&(profile->callprof),
                                    ithread*ithread, // calls
                                    137*ithread);
   }

   for (int ithread=0; ithread<nthreads; ithread++) {
      fprintf(stdout, "Thread: %d ", ithread);
      profile_t *profile = stacktree.stacks[0].profiling.profiles+ithread;
      vftr_print_callprofiling(stdout, profile->callprof);
   }

   vftr_stacktree_free(&stacktree);
   vftr_environment_free(&environment);
#ifdef _MPI
   PMPI_Finalize();
#endif

   FINALIZE_SELF_PROF_VFTRACE;
   return 0;
}
