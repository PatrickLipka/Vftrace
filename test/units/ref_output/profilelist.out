Thread: 0 calls: 0, cycles: 0, time(incl/excl): 0/0
Thread: 1 calls: 1, cycles: 42, time(incl/excl): 137/0
Thread: 2 calls: 4, cycles: 84, time(incl/excl): 274/0
Thread: 3 calls: 9, cycles: 126, time(incl/excl): 411/0
Thread: 4 calls: 16, cycles: 168, time(incl/excl): 548/0
Thread: 5 calls: 25, cycles: 210, time(incl/excl): 685/0
Thread: 6 calls: 36, cycles: 252, time(incl/excl): 822/0
Thread: 7 calls: 49, cycles: 294, time(incl/excl): 959/0
Thread: 8 calls: 64, cycles: 336, time(incl/excl): 1096/0
Thread: 9 calls: 81, cycles: 378, time(incl/excl): 1233/0
